neuraxle/metaopt/auto_ml.py

```diff
@@ -0,0 +1,358 @@
+        return all_hyperparams, all_scores
+
+    def save_score_for_trial(self, hyperparams: HyperparameterSamples, score: float):
+        """
+        Save hyperparams, and score for a trial.
+
+        :return: (hyperparams, scores)
+        """
+        self._save_trial_json(hyperparams, score)
+
+    def _save_trial_json(self, hyperparams, score):
+        """
+        Save trial json file.
+
+        :return: (hyperparams, scores)
+        """
+        hp_dict = hyperparams.to_flat_as_dict_primitive()
+        current_hyperparameters_hash = self._get_trial_hash(hp_dict)
+        with open(os.path.join(self.cache_folder, current_hyperparameters_hash) + '.json', 'w+') as outfile:
+            json.dump({
+                'hyperparams': hp_dict,
+                'score': score
+            }, outfile)
+
+
+class AutoMLStrategyMixin:
+    """
+    Base class for Automatic Machine Learning strategies.
+    Implement your own custom intelligent search of hyperparameters to get most accurate predictive models.
+
+    .. seealso::
+        :class:`BaseCrossValidation`,
+        :class:`HyperparameterSamples`,
+        :class:`HyperparameterSpace`,
+        :class:`MetaStepMixin`,
+        :class:`BaseStep`
+    """
+
+    def __init__(self, validation_technique: BaseCrossValidationWrapper = None, higher_score_is_better=True):
+        if validation_technique is None:
+            validation_technique = KFoldCrossValidationWrapper()
+        self.validation_technique = validation_technique
+        self.higher_score_is_better = higher_score_is_better
+
+    @abstractmethod
+    def find_next_best_hyperparams(self, trials_data_container: 'AutoMLDataContainer') -> HyperparameterSamples:
+        """
+        Find the next best hyperparams using previous trials.
+
+        :param trials_data_container: trials data container
+        :type trials_data_container: AutoMLDataContainer
+        :return: next best hyperparams
+        :rtype: HyperparameterSamples
+        """
+        raise NotImplementedError()
+
+    def fit_transform(self, data_inputs, expected_outputs=None) -> ('AutoMLStrategyMixin', float):
+        """
+        Fit cross validation with wrapped step, and return the score.
+
+        :param data_inputs: data inputs
+        :param expected_outputs: expected outputs to fit on
+        :return: self, step score
+        :rtype: (AutoMLStrategyMixin, float)
+        """
+        step = copy.copy(self.wrapped)
+        step: BaseCrossValidationWrapper = copy.copy(self.validation_technique).set_step(step)
+        step = step.fit(data_inputs, expected_outputs)
+
+        return self, step.get_score()
+
+    def is_higher_score_better(self):
+        return self.higher_score_is_better
+
+    def fit(self, data_inputs, expected_outputs=None):
+        return self
+
+    def transform(self, data_inputs, expected_outputs=None):
+        return data_inputs
+
+
+class RandomSearchBaseAutoMLStrategy(AutoMLStrategyMixin, MetaStepMixin, BaseStep):
```

Comment:
`class RandomSearchAlgorithm` would sound better. I'd also inherit from only 1 object that would be the `BaseAutoMLStrategyAlgorithm`

Reply:
did an awesome refactor, and updated the description of the pr. check it out @guillaume-chevalier